tests/e2e/pipeline.go

```diff
@@ -488,21 +510,73 @@ func updatePipelineRollout(name string, newSpec numaflowv1.PipelineSpec, expecte
 	if expectedFinalPhase == numaflowv1.PipelinePhaseRunning {
-		verifyPipelineRolloutHealthy(name)
+		VerifyPipelineRolloutHealthy(name)
 	}
-
-	verifyInProgressStrategy(name, apiv1.UpgradeStrategyNoOp)
-
-	if expectedFinalPhase == numaflowv1.PipelinePhasePaused {
-		verifyPipelinePaused(Namespace, name)
+	// slow pausing case
+	if expectedFinalPhase == numaflowv1.PipelinePhasePausing && UpgradeStrategy == config.PPNDStrategyID {
+		VerifyInProgressStrategy(name, apiv1.UpgradeStrategyPPND)
 	} else {
-		verifyPipelineRunning(Namespace, name, overrideSourceVertexReplicas)
+		VerifyInProgressStrategy(name, apiv1.UpgradeStrategyNoOp)
+	}
+
+	switch expectedFinalPhase {
+	case numaflowv1.PipelinePhasePausing:
+		VerifyPipelinePausing(Namespace, name)
+	case numaflowv1.PipelinePhasePaused:
+		VerifyPipelinePaused(Namespace, name)
+	case numaflowv1.PipelinePhaseFailed:
+		VerifyPipelineFailed(Namespace, name)
 	}
+
+}
+
+func VerifyPipelineStaysPaused(pipelineRolloutName string) {
+	Document("verifying Pipeline stays in paused or otherwise pausing")
+	Consistently(func() bool {
+		rollout, _ := pipelineRolloutClient.Get(ctx, pipelineRolloutName, metav1.GetOptions{})
+		_, _, retrievedPipelineStatus, err := getPipelineSpecAndStatus(Namespace, pipelineRolloutName)
+		if err != nil {
+			return false
+		}
+		return getRolloutConditionStatus(rollout.Status.Conditions, apiv1.ConditionPipelinePausingOrPaused) == metav1.ConditionTrue &&
+			(retrievedPipelineStatus.Phase == numaflowv1.PipelinePhasePaused || retrievedPipelineStatus.Phase == numaflowv1.PipelinePhasePausing)
+	}, 10*time.Second, testPollingInterval).Should(BeTrue())
+
+	VerifyInProgressStrategy(pipelineRolloutName, apiv1.UpgradeStrategyNoOp)
+
+	pipeline, err := getPipeline(Namespace, pipelineRolloutName)
+	Expect(err).ShouldNot(HaveOccurred())
+	verifyPodsRunning(Namespace, 0, getVertexLabelSelector(pipeline.GetName()))
+}
+
+func CreateFailedPipelineRollout(name, namespace string, spec numaflowv1.PipelineSpec) {
+
+	pipelineRolloutSpec := createPipelineRolloutSpec(name, namespace, spec)
+	_, err := pipelineRolloutClient.Create(ctx, pipelineRolloutSpec, metav1.CreateOptions{})
+	Expect(err).ShouldNot(HaveOccurred())
+
+	Document("Verifying that the PipelineRollout was created")
+	Eventually(func() error {
+		_, err := pipelineRolloutClient.Get(ctx, name, metav1.GetOptions{})
+		return err
+	}, testTimeout, testPollingInterval).Should(Succeed())
+
+	Document("Verifying that the Pipeline was created")
+	VerifyPipelineSpec(namespace, name, func(retrievedPipelineSpec numaflowv1.PipelineSpec) bool {
+		return len(spec.Vertices) == len(retrievedPipelineSpec.Vertices) // TODO: make less kludgey
+		//return reflect.DeepEqual(pipelineSpec, retrievedPipelineSpec) // this may have had some false negatives due to "lifecycle" field maybe, or null values in one
+	})
+
+	VerifyPipelineRolloutDeployed(name)
+	VerifyInProgressStrategy(name, apiv1.UpgradeStrategyNoOp)
```

Comment:
How is this function different from `CreatePipelineRollout()`?

Reply:
The only difference is we don't check if the Pipeline is running and the vertices are up.

Reply:
We could probably safely add another `failed` boolean to this to handle that case and be sure we don't have multiple repeating functions.

Reply:
Added `failed` boolean.

Reply:
sounds good